Step5 - 로또(수동)

lewisseo91 opened the pull request:
안녕하세요!
step5 - 로또(수동)까지 완료하고 PR 보냅니다!
많은 좋은 피드백 주신 덕분에 어떻게 PR을 보낼 수가 있었습니다.
정말 감사합니다!
잘 부탁드립니다~!

lewisseo91 committed:
style(FixtureTest): 테스트 lottoNumberOf 줄 바꿈 수정

FixtureTest 한 줄에 점 하나만 찍는다. -> 규칙 지키기! 잊지 말기!

lewisseo91 committed:
style(LottoRankingStatus): LottoRankingStatus 의미있는 상수명으로 변경

LottoRankingStatus 의미있는 상수명으로 변경

lewisseo91 committed:
feat(BuyAmountTest): manualAmount 추가, 구매 개수 예외 추가

변수 manualAmount 추가
구매 개수 예외 추가
- 총 구매 개수보다 수동 구매 개수를 더 크게 구매할 시 exception 발생 테스트

lewisseo91 committed:
feat(BuyAmount): manualAmount 추가, 구매 개수 예외 확인 추가

변수 manualAmount 추가
구매 개수 예외 확인 추가
- validateNotExceedAmountSize 메서드 추가

lewisseo91 committed:
feat(LottoNumbersTest): LottoNumbersTest 수동 입력 오류 추가

LottoNumbersTest 수동 입력 오류 추가
- generateManualNullInputNumbersException 테스트 추가
- 이에 따른 throws Exception 추가
LottoNumbersGroupTest throws Exception 추가
- LottoNumbers 받아오는 부분 모두 throws Exception 추가

lewisseo91 committed:
feat(LottoNumbersGroup): LottoNumbers 수동 입력 추가, throws Exception 추가

LottoNumbersGroup 수동 입력 추가
- 외부에서 만든 로또 수동 숫자 입력 추가 후 자동 숫자 생성
- LottoNumbers 수동 입력 null 에 따른 throws Exception 추가
LottoNumbers throws Exception 추가
- LottoNumbers 수동 입력 null 시 Optional 사용 throw
- LottoNumbers 수동 입력 null 에 따른 throws Exception 추가
LottoApplication, LottoController throws Exception 추가

lewisseo91 committed:
feat(LottoController): 수동 입력 받아오는 부분 추가

수동 입력 받아오는 부분 추가
- getManualBuyAmount 수동 구매 숫자 받아오는 메서드 추가
- getManualLottoNumbers 수동 입력 로또 숫자 받아오는 메서드 추가

lewisseo91 committed:
feat(LottoMessage): 수동 입력 후 구매 메시지 변경

수동 입력 후 구매 메시지 변경
- n개 구매 -> 수동 n장, 자동 n개 구매 메시지 표시

lewisseo91 committed:
refactor(LottoNumbers): 수동 입력 후 Optional 리팩토링

수동 입력 후 Optional 리팩토링

lewisseo91 committed:
style(LottoNumbersGroup): LottoNumbersGroup 변수 naming 변경

LottoNumbersGroup 변수 naming 변경
- 여러 개의 LottoNumbers 인데도 불구하고 같은 Numbers 이름으로 끝나서 변경

lewisseo91 committed:
style(LottoRankingStatus): LottoRankingStatus NONE 범위 3 이하로만 변경

LottoRankingStatus NONE 범위 3 이하로만 변경
- 6 초과는 WrongLottoRankingStatusException 발생
LottoMessage 에 WrongLottoRankingStatusException 관련 메시지 추가
WrongLottoRankingStatusException 6 초과 로또 등수 예외 추가

lewisseo91 commented:
안녕하세요! 현재 PR 제꺼를 보고 복습하다가 
로또 피드백 반영이 안된 부분과 Exception 설정 시에 이상하게 한 것이 있어서 수정 중입니다.
혹시나 보고 계시면 리뷰 잠깐 멈춰주시면 정말 감사하겠습니다!

lewisseo91 commented:
> 안녕하세요! 현재 PR 제꺼를 보고 복습하다가 로또 피드백 반영이 안된 부분과 Exception 설정 시에 이상하게 한 것이 있어서 수정 중입니다. 혹시나 보고 계시면 리뷰 잠깐 멈춰주시면 정말 감사하겠습니다!

아이고 제가 실수로 닫아버렸네요 
수정한 다음에 제가 재요청 하겠습니다!

lewisseo91 committed:
fix(WrongLottoNumbersInputException): WrongLottoNumbersInputException 상속 클래스 수정

Exception -> RuntimeException
throws Exception 모두 제거

lewisseo91 committed:
feat(PrizeLottoNumbersTest): 당첨 로또 객체 생성

당첨 로또 결과 테스트 추가
당첨 로또 객체 오류 테스트 추가

lewisseo91 committed:
feat(PrizeLottoNumbers): 당첨 로또 객체 생성

PrizeLottoNumbers
당첨 로또 객체 추가
당첨 로또 보너스 숫자 포함 예외 추가
LottoResult
equals 추가
LottoNumbersGroup, LottoNumbersGroupTest
기존 getLottoResult 메서드 PrizeLottoNumbers 로 이동 후 반영
LottoMessage
보너스 숫자 포함 예외 메시지 추가
LottoController
기존 getLottoResult 메서드 PrizeLottoNumbers 로 이동 후 반영

lewisseo91 committed:
feat(LottoNumber): lottoNumber 팩토리 메서드 생성, 인스턴스 캐싱

new LottoNumber -> LottoNumber.of 로 변경
validate static 으로 변경 필요한 메서드들 변경

lewisseo91 committed:
feat(LottoNumber): final 키워드 추가

final 키워드 추가

nokchax commented on src/main/java/lotto/domain/BuyAmount.java:
```diff
@@ -7,10 +7,17 @@
     public static final int LOTTO_TICKET_PRICE = 1000;
 
     private final int amount;
+    private final int manualAmount;
```
amount가 하나 더 추가 되었네요?
금액에 관련 된 변수이니 이걸 돈이라는 클래스로 래핑해보는건 어떨까요?

lewisseo91 replied:
> amount가 하나 더 추가 되었네요? 금액에 관련 된 변수이니 이걸 돈이라는 클래스로 래핑해보는건 어떨까요?

리뷰 감사합니다~!
아하 수동 입력을 받고 해당 입력을 그대로 보관한다는 생각을 했는데 감사합니다! 
돈으로 wrapping 해 보겠습니다.

nokchax commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -3,46 +3,59 @@
 public class LottoNumber implements Comparable<LottoNumber> {
     public static final int MIN_BOUND = 1;
     public static final int MAX_BOUND = 45;
+    private static final Map<Integer, LottoNumber> lottoNumbers = createLottoNumbers();
```
캐싱 처리 👍 
```java
private static final Map<Integer, LottoNumber> lottoNumbers = 
   IntStream.rangeClosed(MIN_BOUND, MAX_BOUND)
                    .mapToObj(LottoNumber::new)
                    .collect(Collectors.toMap(lottoNumber -> lottoNumber.number, Function.identity()));
```
스트림을 사용하여 이렇게 선언과 동시에 초기화 할 수 있습니다 :)

lewisseo91 replied:
> 캐싱 처리 👍
> 
> ```java
> private static final Map<Integer, LottoNumber> lottoNumbers = 
>    IntStream.rangeClosed(MIN_BOUND, MAX_BOUND)
>                     .mapToObj(LottoNumber::new)
>                     .collect(Collectors.toMap(lottoNumber -> lottoNumber.number, Function.identity()));
> ```
> 
> 스트림을 사용하여 이렇게 선언과 동시에 초기화 할 수 있습니다 :)

와 감사합니다 스트림에 대해 좀 더 조사해 보겠습니다~!

nokchax commented on src/main/java/lotto/domain/LottoRankingStatus.java:
```diff
@@ -1,13 +1,16 @@
-    MATCH5(5, false, 1_500_000),
-    MATCH5BONUS1(5, true, 30_000_000),
-    MATCH6(6, false, 2_000_000_000),
+    FIFTH(3, false, 5_000),
+    FOURTH(4, false, 50_000),
+    THIRD(5, false, 1_500_000),
+    SECOND(5, true, 30_000_000),
+    FIRST(6, false, 2_000_000_000),
```
👍

nokchax reviewed (changes requested):
안녕하세요, 동현님!
기존에 설계가 잘 되어있던 덕에  마지막 단계도 큰 변경없이 구현 완료하셨네요 👍 
한가지 개선할 점이 생겨서 코멘트 드렸는데요
한 번 확인후에 재요청 부탁드립니다 :)

lewisseo91 committed:
style(PrizeLottoNumbers): PrizeLottoNumbersTest displayName 추가

PrizeLottoNumbersTest displayName 추가

lewisseo91 committed:
feat(MoneyTest): Money 테스트 추가

수동 입력 돈이 있을 때 초과 확인 테스트 추가

lewisseo91 committed:
feat(Money): Money 추가

구입 금액, 수동 금액(수동 로또 구매 횟수 * 최소 금액) 저장하는 Money 클래스 추가
getAmount -> getTotalAmount 이름 변경

lewisseo91 commented:
좋은 피드백 감사합니다~!
Money 클래스 wrapping 후에 PR 보냅니다~
금액으로 wrapping을 시도 했는데 혹시 관점에 따라서는 Amount로 wrapping 해도 상관 없는 지 궁금합니다!
현재 로직 내에서 금액은 구매 이후에 갯수 셀 때만 들어가는 것 같아서요!

lewisseo91 committed:
refactor(LottoNumber): LottoNumber IntStream 으로 number 생성 추가

IntStream 으로 number 생성

lewisseo91 commented:
좋은 피드백 감사합니다~!
Money 클래스 wrapping 후에 PR 보냅니다~
>~~금액으로 wrapping을 시도 했는데 혹시 관점에 따라서는 Amount로 wrapping 해도 상관 없는 지 궁금합니다!~~
>~~현재 로직 내에서 금액은 구매 이후에 갯수 셀 때만 들어가는 것 같아서요!~~

-> 생각해보니 수익율 구할 때 필요한 걸 다시 보고 알았습니다 

리뷰 감사합니다!

nokchax commented on src/main/java/lotto/domain/BuyAmount.java:
```diff
@@ -6,11 +6,18 @@
 public class BuyAmount {
     public static final int LOTTO_TICKET_PRICE = 1000;
```
price 또한 money라고 봐야하지 않을까요?

lewisseo91 replied:
아하 맞습니다! 
상수에 들어가는 값은 원시값만 생각하고 있었습니다. 
제 생각의 틀이 너무 좁았네요 감사합니다~!

아니라면 아예 money 안 쪽에서 price를 컨트롤 하는 것으로 생각해야겠네요 감사합니다!

nokchax reviewed (changes requested):
동현님 하나만 더 수정해 주셨으면해서 코멘트남겼습니다~!
확인 부탁드려요

lewisseo91 committed:
refactor(Money): BuyAmount 에 있던 price -> Money 로 이동

BuyAmount 에 있던 price -> Money 로 이동
기존 BuyAmount에서 쓰던 LOTTO_TICKET_PRICE는 Money.LOTTO_TICKET_PRICE 로 변경

lewisseo91 commented:
안녕하세요~
매번 리뷰 해 주셔서 정말 고맙습니다! 🙇 

일단 제가 price를 money에 상수를 옮기고 나서 생각을 해 보았는데 궁금한 점이 생겼습니다.

원래 제가 기준을 로또 구매 or 생성 개수를 기준으로 생각하고 TDD를 진행하였고 여기까지 왔습니다만
Money로 wrapping을 완료하고 또한 LOTTO_TICKET_PRICE를 옮기고 나니
BuyAmount라는 구매 개수를 생각하며 만들었던 클래스의 멤버들이 모두 가격과 관련된 Money를 가지고 있는 아이러니를 겪고 있다고 생각하게 되었습니다.
구매 개수지만 총 구매 가격과 수동 구매 가격을 멤버로 가지는 것이 객체를 잘 설명하고 있다고 할 수 있을까요?
아니라면 다시 가격을 기준으로 리팩토링하는 것이 좋을까요?

lewisseo91 committed:
refactor(Money): manualMoney 를 표현하고 싶어서 정적 팩토리 메서드로 변경

manualMoney 를 표현하고 싶어서 정적 팩토리 메서드로 변경
LOTTO_TICKET_PRICE -> LOTTO_BASE_MONEY 로 변경 후 BuyAmount로 복귀
isExceedMoney 문맥 설명상 왼쪽 오른쪽이 바뀐 것 같아서 자리 바꿈

lewisseo91 committed:
chore(Money): 줄 바꿈 정리

줄 바꿈 정리

lewisseo91 commented:
리뷰 해 주셔서 정말 감사합니다!
한번 더 푸시하고 요청 보냅니다~!
이번에 manualMoney를 적용하기 위해 정적 팩토리 메서드를 이용해 보았습니다.
잘 부탁드립니다~!!

아래는 변수 이름만 변경하고 commit 한 코드입니다~!

lewisseo91 committed:
style(Money): manualMoney -> baseMoney 이름 변경

manualMoney -> baseMoney 이름 변경

nokchax commented:
지금 와서보니 제가 착각한 부분이 있었네요 ㅠ
이번 pr 최초 요청시 amount 와 manualAmount가 금액을 표현하는 걸로 착각했었습니다.
괜히 저때문에 불필요한 수정을 하게 됐네요 죄송합니다. 🙇‍♂️ 

```
BuyAmount라는 구매 개수를 생각하며 만들었던 클래스의 멤버들이 모두 가격과 관련된 Money를 가지고 있는 아이러니를 겪고 있다고 생각하게 되었습니다.
구매 개수지만 총 구매 가격과 수동 구매 가격을 멤버로 가지는 것이 객체를 잘 설명하고 있다고 할 수 있을까요?
```
음 아니요 제가 혼란을 드려 현재는 구입개수 라는 클래스에 돈과 관련된 Money 객체가 들어가게 되었고,
그래서 클래스 명에 이질감이 생겨버렸네요. 본래의 구입 개수, 수동 구입 개수 그대로 두셨다면 이상할게 없습니다.

만약에 지금의 Money 멤버를 유지하신다면, 클래스 명을 구매 정보(e.g. PurchaseInfo)와 같은 더 어울릴만한 클래스 명으로 변경이 필요할 것 같습니다. ㅠ 다시 한 번 죄송합니다.

lewisseo91 commented:
> 지금 와서보니 제가 착각한 부분이 있었네요 ㅠ 이번 pr 최초 요청시 amount 와 manualAmount가 금액을 표현하는 걸로 착각했었습니다. 괜히 저때문에 불필요한 수정을 하게 됐네요 죄송합니다. 🙇‍♂️
> 
> ```
> BuyAmount라는 구매 개수를 생각하며 만들었던 클래스의 멤버들이 모두 가격과 관련된 Money를 가지고 있는 아이러니를 겪고 있다고 생각하게 되었습니다.
> 구매 개수지만 총 구매 가격과 수동 구매 가격을 멤버로 가지는 것이 객체를 잘 설명하고 있다고 할 수 있을까요?
> ```
> 
> 음 아니요 제가 혼란을 드려 현재는 구입개수 라는 클래스에 돈과 관련된 Money 객체가 들어가게 되었고, 그래서 클래스 명에 이질감이 생겨버렸네요. 본래의 구입 개수, 수동 구입 개수 그대로 두셨다면 이상할게 없습니다.
> 
> 만약에 지금의 Money 멤버를 유지하신다면, 클래스 명을 구매 정보(e.g. PurchaseInfo)와 같은 더 어울릴만한 클래스 명으로 변경이 필요할 것 같습니다. ㅠ 다시 한 번 죄송합니다.

아닙니다~! 제가 오해의 여지가 있는 설계를 했다고 생각합니다~!
 buyAmount인데 생성자에서는 buyPrice를 받았기 때문에 저도 사과드립니다.
 아 그리고 buyAmount 라는 말이 정말 모호하기도 하기 때문에 좀 더 명확한 용어를 사용해야 할 것 같다고 반성하고 있습니다. 혼란을 드려서 죄송해요!
 그리고 리팩토링이 정말 즐거워서 오히려 좋습니닼ㅋㅋㅋㅋ
 그럼 수정하고 다시 요청 드리겠습니다~

nokchax reviewed (approved):
안녕하세요, 동현님 제가 착각하는 바람에 괜히 더 코드를 수정하게 됐네요 ㅠ 죄송합니다.
그동안 부족한 리뷰 따라와 주셔서 감사합니다. 마지막 단계 머지할게요 💯 
리뷰 진행하는 동안 커뮤니케이션을 자주 할 수 있어서 좋았습니다 :)
동현님도 다른 분들의 코드를 보면서 공부하시는 것도 도움이 많이 되실거예요.
앞으로 남은 과제도 파이팅입니다 👍

nokchax commented:
머지는 추가적으로 코드 수정하시면 진행하도록 할게요 :)

lewisseo91 commented:
> 안녕하세요, 동현님 제가 착각하는 바람에 괜히 더 코드를 수정하게 됐네요 ㅠ 죄송합니다. 그동안 부족한 리뷰 따라와 주셔서 감사합니다. 마지막 단계 머지할게요 💯 리뷰 진행하는 동안 커뮤니케이션을 자주 할 수 있어서 좋았습니다 :) 동현님도 다른 분들의 코드를 보면서 공부하시는 것도 도움이 많이 되실거예요. 앞으로 남은 과제도 파이팅입니다 👍

정말 좋은 피드백들 주셔서 감사드립니다 ^^!!
다른 분들 코드를 보는 것도 정말 즐겁더라구요. 
'나라면 어땠을까', '아하 이런 부분은 이렇게 해야하는 구나' 하고 깨닫는 부분에 감동을 느꼈습니다.
좋은 피드백들 많이 주셔서 감사합니다 바로 커밋 올리겠습니다 😄

lewisseo91 committed:
refactor(PurchaseInfo): buyAmount -> PurchaseInfo 로 네이밍 변경

오해의 여지가 있는 네이밍 피할 것!
BuyAmount (구매량 -> 구매 수량, 구매 가격 둘 다 표현 가능하기 떄문에 모호)
-> PurchaseInfo (리뷰어님 추천 네이밍 아주 좋음 혼자 1시간 고민했는데 이런 네이밍이 안 나왔는데 부럽당..)
manualAmount (수동량 -> 수동 구매량, 수동 가격 둘 다 표현 가능하기 떄문에 모호)
-> manualCount 로 변경

lewisseo91 commented:
nokchax 리뷰어님 덕분에 미션 기간동안에 많은 즐거움과 깨달음을 얻을 수 있었습니다. 😄 
감사합니다! 🙇 
리뷰 잘 부탁드립니다~!

nokchax commented on src/main/java/lotto/domain/PurchaseInfo.java:
```diff
@@ -3,19 +3,19 @@
     }
 
-    public BuyAmount(int buyPrice, int manualAmount) {
+    public PurchaseInfo(int buyPrice, int manualCount) {
```
변수명 네이밍은 언제나 어려운 거 같아요 👍

lewisseo91 replied:
맞습니다 ㅠ 좋은 이름 제시 해 주셔서 감사합니다~!
다른게 떠오르지 않아 쓰윽 배꼈습니다..

nokchax commented on src/main/java/lotto/ui/LottoController.java:
```diff
@@ -13,18 +13,18 @@ public void startGame() {
         int manualBuyAmount = getManualBuyAmount();
-        BuyAmount buyAmount = new BuyAmount(buyPrice, manualBuyAmount);
-        LottoNumbersGroup lottoNumbersGroup = new LottoNumbersGroup(buyAmount, getManualLottoNumbers(buyAmount));
+        PurchaseInfo purchaseInfo = new PurchaseInfo(buyPrice, manualBuyAmount);
```
클래스명이 수정되면서 더 잘 읽히는거 같습니다 :)

nokchax reviewed (approved):
고생하셨습니다! 머지할게요 :)
